examples/ReinterpretLog.hs

```diff
@@ -0,0 +1,159 @@
+-- This example shows how to reinterpret a simple, first-order "logging" effect,
+-- in terms of itself, in order to change the type of the values it logs.
+--
+-- * First, we will define a structured log message type, which is the type our
+--   application prefers to log in.
+--
+-- * Next, we will define a logging carrier that prints strings to stdout.
+--
+-- * Finally, we will bridge the two with an effect carrier that reinterpretes
+--   structured log messages as strings.
+
+
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE InstanceSigs               #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+module ReinterpretLog where
+
+import Control.Effect
+import Control.Effect.Carrier
+import Control.Effect.Sum
+import Control.Monad.IO.Class     (MonadIO(..))
+import Control.Monad.Trans.Reader (ReaderT(..))
+import Data.Coerce                (coerce)
+import Data.Function              ((&))
+import Data.Kind                  (Type)
+import Prelude                    hiding (log)
+
+
+--------------------------------------------------------------------------------
+-- The application
+--------------------------------------------------------------------------------
+
+-- Our structured log message.
+data Message
+  = Debug String
+  | Info String
+
+-- Render a structured log message as a string.
+renderLogMessage :: Message -> String
+renderLogMessage = \case
+  Debug message -> "[debug] " ++ message
+  Info  message -> "[info] "  ++ message
+
+-- The application: it logs two messages, then quits.
+application :: (Carrier sig m, Member (Log Message) sig) => m ()
+application = do
+  log (Debug "debug message")
+  log (Info "info message")
+
+-- The application runner. Interpret the application by:
+--
+-- * Reinterpreting 'Log Message' effects as 'Log String' effects.
+-- * Interpreting 'Log String' effects by printing to stdout.
+runApplication :: IO ()
+runApplication =
+  application
+    & reinterpretLog renderLogMessage
+    & runLogStdout
+    & runM
+
+
+--------------------------------------------------------------------------------
+-- The logging effect
+--------------------------------------------------------------------------------
+
+-- Log an 'a', then continue with 'k'.
+data Log :: Type -> (Type -> Type) -> Type -> Type where
+  Log :: a -> k -> Log a m k
+  deriving stock (Functor)
```

Comment:
Would it be better (for the purpose of pedagogy) to write this without GADT syntax? Just thinking that this requires readers to already understand one more thing than they would otherwise have to.

Reply:
Will do!